README.md
# Passwordless-RedisStore

This module provides token storage for [Passwordless](https://github.com/florianheinemann/passwordless), a node.js module for express that allows website authentication without password using verification through email or other means. Visit the project's [website](https://passwordless.net) for more details.

Tokens are stored in a Redis database and are hashed and salted using [bcrypt](https://github.com/ncb000gt/node.bcrypt.js/).

## Usage

First, install the module:

`$ npm install passwordless-redisstore --save`

Afterwards, follow the guide for [Passwordless](https://github.com/florianheinemann/passwordless). A typical implementation may look like this:

```javascript
var passwordless = require('passwordless');
var RedisStore = require('passwordless-redisstore');

passwordless.init(new RedisStore(6379, '127.0.0.1'));

passwordless.addDelivery(
    function(tokenToSend, uidToSend, recipient, callback) {
        // Send out a token
    });
    
app.use(passwordless.sessionSupport());
app.use(passwordless.acceptToken());
```

## Initialization

```javascript
new RedisStore([port], [host], [options]);
```
* **[port]:** *(Number)* Optional. Port of your Redis server. Defaults to: 6379
* **[host]:** *(String)* Optional. Your Redis server. Defaults to: '127.0.0.1'
* **[options]:** *(Object)* Optional. This can include options of the node.js Redis client as described in the [docs](https://github.com/mranney/node_redis) and the ones described below combined in one object as shown in the example

Example:
```javascript
passwordless.init(new RedisStore(6379, '127.0.0.1', {
	// option of the node.js redis client
    auth_pass: 'password',
    // options of RedisStore
    redisstore: {
        database: 15,
        tokenkey: 'token:',
        rounds: 12
    }
}));
```

### Options
* **[redisstore.database]:** *(Number)* Optional. Database to be used. Defaults to: 0
* **[redisstore.tokenkey]:** *(String)* Optional. Keys to be used. UIDs will be appended. Defaults to: 'pwdless:UID'
* **[redisstore.rounds]:** *(Number)* Optional. Number of bccrypt rounds to be used. Defaults to: 12

## Hash and salt
As the tokens are equivalent to passwords (even though only for a limited time) they have to be protected in the same way. passwordless-redisstore uses [bcrypt](https://github.com/ncb000gt/node.bcrypt.js/) with automatically created random salts. To generate the salt 12 rounds are used. This number can be overridden with a larger integer in the options.

## Tests

`$ npm test`

## License

[MIT License](http://opensource.org/licenses/MIT)

## Author
Florian Heinemann [@thesumofall](http://twitter.com/thesumofall/)


CHANGELOG.md
# 1.0.2 (2016-03-25)
Bugfixes:
- FIX storeOrUpdate to return an empty string as referrer if it was passed as null (in line with spec)
- UPDATE dependencies

# 1.0.1 (2015-08-22)
Bugfixes:
- FIX API documentation which incorrectly referred to MongoDB

# 1.0.0 (2015-07-13)
Features:
- UPDATE of all requires incl. Redis

Bugfixes:
- FIX Throws were not detected properly
- FIX Empty initialization of RedisStore() caused issues

# 0.0.1 (2014-06-28)
- Initial release

test/redisstore.js
'use strict';

var expect = require('chai').expect;
var uuid = require('node-uuid');
var chance = new require('chance')();

var RedisStore = require('../');
var TokenStore = require('passwordless-tokenstore');

var redis = require("redis");
var client = null;
var db = 15;

var standardTests = require('passwordless-tokenstore-test');

function TokenStoreFactory() {
	return new RedisStore(null, null, { redisstore: {database: db }} );
}

var beforeEachTest = function(done) {
	if(!client) {
		client = redis.createClient();
	}

	client.select(db, function(err, res) {
		done(err);
	})
}

var afterEachTest = function(done) {
	client.flushdb(function(err, res) {
		done(err);
	})
}

// Call all standard tests
standardTests(TokenStoreFactory, beforeEachTest, afterEachTest);

describe('Specific tests', function() {

	beforeEach(function(done) {
		beforeEachTest(done);
	})

	afterEach(function(done) {
		afterEachTest(done);
	})

	it('should allow the instantiation with an empty constructor', function () {
		expect(function() { new RedisStore() }).to.not.throw();
	})

	it('should allow the instantiation with host and port but no options', function () {
		expect(function() { new RedisStore(6379, '127.0.0.1') }).to.not.throw();
	})

	it('should allow the instantiation with a number passed as DB selector', function () {
		expect(function() { new RedisStore(null, null, {redisstore : { database: 0}}) }).to.not.throw();
	})

	it('should allow proper instantiation', function () {
		expect(function() { TokenStoreFactory() }).to.not.throw();
	})

	it('should not allow the instantiation with a DB selector that is not a number', function () {
		expect(function() { new RedisStore(null, null, {redisstore : { database: 'test'}}) }).to.throw(Error);
	})

	it('should default to 0 as database', function(done) {
		var store = new RedisStore();

		var user = chance.email();
		store.storeOrUpdate(uuid.v4(), user, 
			1000*60, 'http://' + chance.domain() + '/page.html', 
			function() {
				client.select(0, function(err, res) {
					expect(err).to.not.exist;
					client.hgetall('pwdless:' + user, function(err, obj) {
						expect(err).to.not.exist;
						expect(obj).to.exist;
						client.del('pwdless:' + user, function(err, dels) {
							expect(err).to.not.exist;
							expect(dels).to.equal(1);
							done();
						})
					})
				})
			});
	});

	it('should change name of token key based on "redisstore.tokenkey"', function(done) {
		var store = new RedisStore(null, null, { redisstore : { tokenkey: 'another_name_', database: db }});

		var user = chance.email();
		store.storeOrUpdate(uuid.v4(), user, 
			1000*60, 'http://' + chance.domain() + '/page.html', 
			function() {
				client.hgetall('another_name_' + user, function(err, obj) {
					expect(err).to.not.exist;
					expect(obj).to.exist;
					done();
				})
			});		
	});

	it('should default to "pwdless:" as token key', function(done) {
		var store = TokenStoreFactory();
		var user = chance.email();
		store.storeOrUpdate(uuid.v4(), user, 
			1000*60, 'http://' + chance.domain() + '/page.html', 
			function() {
				client.hgetall('pwdless:' + user, function(err, obj) {
					expect(err).to.not.exist;
					expect(obj).to.exist;
					done();
				})
			});
	});

	it('should store tokens only in their hashed form', function(done) {
		var store = TokenStoreFactory();
		var user = chance.email();
		var token = uuid.v4();
		store.storeOrUpdate(token, user, 
			1000*60, 'http://' + chance.domain() + '/page.html', 
			function() {
				client.hgetall('pwdless:' + user, function(err, obj) {
					expect(err).to.not.exist;
					expect(obj).to.exist;
					expect(obj.token).to.exist;
					expect(obj.token).to.not.equal(token);
					done();
				})
			});
	});

	it('should store tokens not only hashed but also salted', function(done) {
		var store = TokenStoreFactory();
		var user = chance.email();
		var token = uuid.v4();
		var hashedToken1;
		store.storeOrUpdate(token, user, 
			1000*60, 'http://' + chance.domain() + '/page.html', 
			function() {
				client.hgetall('pwdless:' + user, function(err, obj) {
					expect(err).to.not.exist;
					expect(obj).to.exist;
					expect(obj.token).to.exist;
					hashedToken1 = obj.token;
					store.storeOrUpdate(token, user, 
						1000*60, 'http://' + chance.domain() + '/page.html', 
						function() {
							client.hgetall('pwdless:' + user, function(err, obj) {
								expect(err).to.not.exist;
								expect(obj).to.exist;
								expect(obj.token).to.exist;
								expect(obj.token).to.not.equal(hashedToken1);
								done();
							});						
						});
				})
			});		
	});
})

lib/redisstore.js
'use strict';

var util = require('util');
var bcrypt = require('bcrypt');
var TokenStore = require('passwordless-tokenstore');
var redis = require('redis');
var async = require('async');

/**
 * Constructor of RedisStore
 * @param {Number} [port] Port of the Redis server (default: 6379)
 * @param {String} [host] Host, e.g. 127.0.0.1 (default: 127.0.0.1)
 * @param {Object} [options] Combines both the options for the Redis client as well
 * as the options for RedisStore. For the Redis options please refer back to
 * the documentation: https://github.com/NodeRedis/node_redis. RedisStore accepts
 * the following options: (1) { redisstore: { database: Number }} number of the 
 * Redis database to use (default: 0), (2) { redisstore: { tokenkey: String }} the
 * prefix used for the RedisStore entries in the database (default: 'pwdless:')
 * @constructor
 */
function RedisStore(port, host, options) {
	port = port || 6379;
	host = host || '127.0.0.1';
	this._options = options || {};
	this._options.redisstore = this._options.redisstore || {};
	if(this._options.redisstore.database && !isNumber(this._options.redisstore.database)) {
		throw new Error('database has to be a number (if provided at all)');
	} else if(this._options.redisstore.database) {
		this._database = this._options.redisstore.database;
	} else {
		this._database = 0;
	}

	this._tokenKey = this._options.redisstore.tokenkey || 'pwdless:';
	delete this._options.redisstore;

	this._client = redis.createClient(port, host, this._options);
}

util.inherits(RedisStore, TokenStore);

/**
 * Checks if the provided token / user id combination exists and is
 * valid in terms of time-to-live. If yes, the method provides the
 * the stored referrer URL if any.
 * @param  {String}   token to be authenticated
 * @param  {String}   uid Unique identifier of an user
 * @param  {Function} callback in the format (error, valid, referrer).
 * In case of error, error will provide details, valid will be false and
 * referrer will be null. If the token / uid combination was not found
 * found, valid will be false and all else null. Otherwise, valid will
 * be true, referrer will (if provided when the token was stored) the
 * original URL requested and error will be null.
 */
RedisStore.prototype.authenticate = function(token, uid, callback) {
	if(!token || !uid || !callback) {
		throw new Error('TokenStore:authenticate called with invalid parameters');
	}

	var self = this;
	self._select(function(err) {
		if(err) {
			return callback(err, false, null);
		}
		var key = self._tokenKey + uid;
		self._client.hgetall(key, function(err, obj) {
			if(err) {
				return callback(err, false, null);
			}
			else if(!obj) {
				return callback(null, false, null);
			} else if(Date.now() > obj.ttl) {
				callback(null, false, null);
			} else {
				bcrypt.compare(token, obj.token, function(err, res) {
					if(err) {
						callback(err, false, null);
					} else if(res) {
						callback(null, true, obj.origin);
					} else {
						callback(null, false, null);
					}
				});
			}
		});
	})
};

/**
 * Stores a new token / user ID combination or updates the token of an
 * existing user ID if that ID already exists. Hence, a user can only
 * have one valid token at a time
 * @param  {String}   token Token that allows authentication of _uid_
 * @param  {String}   uid Unique identifier of an user
 * @param  {Number}   msToLive Validity of the token in ms
 * @param  {String}   originUrl Originally requested URL or null
 * @param  {Function} callback Called with callback(error) in case of an
 * error or as callback() if the token was successully stored / updated
 */
RedisStore.prototype.storeOrUpdate = function(token, uid, msToLive, originUrl, callback) {
	if(!token || !uid || !msToLive || !callback || !isNumber(msToLive)) {
		throw new Error('TokenStore:storeOrUpdate called with invalid parameters');
	}

	var self = this;
	self._select(function(err) {
		if(err) {
			return callback(err, false, null);
		}
		bcrypt.hash(token, 10, function(err, hashedToken) {
			if(err) {
				return callback(err);
			}

			originUrl = originUrl || "";
			var key = self._tokenKey + uid;
			self._client.hmset(key, {
					token: hashedToken,
					origin: originUrl,
					ttl: (Date.now() + msToLive)
				}, function(err, res) {
					if(!err) {
						msToLive = Math.ceil(msToLive / 1000);
						self._client.expire(key, msToLive, function(err, res) {
							if(err)
								callback(err);
							else
								callback();
						})
					} else {
						callback(err);
					}
			});
		});
	})
}

/**
 * Invalidates and removes a user and the linked token
 * @param  {String}   user ID
 * @param  {Function} callback called with callback(error) in case of an
 * error or as callback() if the uid was successully invalidated
 */
RedisStore.prototype.invalidateUser = function(uid, callback) {
	if(!uid || !callback) {
		throw new Error('TokenStore:invalidateUser called with invalid parameters');
	}

	var self = this;
	self._select(function(err) {
		if(err) {
			return callback(err, false, null);
		}
		var key = self._tokenKey + uid;
		self._client.del(key, function(err) {
			if(err)
				callback(err);
			else
				callback();
		});
	})
}

/**
 * Removes and invalidates all token
 * @param  {Function} callback Called with callback(error) in case of an
 * error or as callback() otherwise
 */
RedisStore.prototype.clear = function(callback) {
	if(!callback) {
		throw new Error('TokenStore:clear called with invalid parameters');
	}

	var self = this;
	self._select(function(err) {
		if(err) {
			return callback(err, false, null);
		}
		var pattern = self._tokenKey + '*';
		self._client.keys(pattern, function(err, matches) {
			if(err) {
				callback(err);
			} else if(matches.length > 0) {
				async.each(matches, function(match, matchCallback) {
					self._client.del(match, matchCallback);
				}, callback);
			} else {
				callback();
			}
		});
	})
}

/**
 * Number of tokens stored (no matter the validity)
 * @param  {Function} callback Called with callback(null, count) in case
 * of success or with callback(error) in case of an error
 */
RedisStore.prototype.length = function(callback) {
	if(!callback) {
		throw new Error('TokenStore:length called with invalid parameters');
	}

	var self = this;
	self._select(function(err) {
		if(err) {
			return callback(err, false, null);
		}
		var pattern = self._tokenKey + '*';
		self._client.keys(pattern, function(err, matches) {
			if(err) {
				callback(err);
			} else {
				callback(null, matches.length);
			}
		});
	})
}

RedisStore.prototype._select = function(callback) {
	var self = this;
	if(self._selected) {
		callback();
	} else {
		self._client.select(self._database, function(err, res) {
			if(err) {
				callback(err);
			} else {
				self._selected = true;
				callback();
			}
		});
	}
}

function isNumber(n) {
	return !isNaN(parseFloat(n)) && isFinite(n);
}

module.exports = RedisStore;
